### PR TITLE
fix(payload): match hyperswitch request parity (string amount + ACH bank_debit fields)

### PR DESCRIPTION
**Pull request by shuklatushar226**
## What

Shadow validation flagged two `payload` request-parity diffs against the hyperswitch reference:

- `payload / bank_debit / bank_debit / authorize` — [juspay/hyperswitch-cloud#16753](https://github.com/juspay/hyperswitch-cloud/issues/16753)
- `payload / card / card / refund` — [juspay/hyperswitch-cloud#16754](https://github.com/juspay/hyperswitch-cloud/issues/16754)

| Field | hyperswitch (reference) | UCS (before) |
|---|---|---|
| request `amount` | string `"60.00"` | number `60.0` |
| `payment_method.bank_account.account_class` | present (`null` when absent) | missing |
| `payment_method.bank_account.account_currency` | `"USD"` | missing |
| `payment_method.billing_address` | present | missing |
| `description` | present | missing |
| `attrs` | filtered metadata | missing |

Method (`POST`) and URL already matched. The `x-merchant-id` header and the random multipart-boundary differences in the report are connector-service shadow infra / cosmetic, not connector-controlled.

## Root cause

1. **Amount type.** The Payload amount converter used `FloatMajorUnit`, whose inner `f64` serde-serializes as a bare JSON number (`60.0`). hyperswitch uses `StringMajorUnitForConnector` → `StringMajorUnit`, a 2-decimal string (`"60.00"`). This drives the `amount` field of every request flow (authorize, refund, mandate, token).
2. **ACH bank_debit body.** `build_payload_bank_account_request_data` omitted `account_class`, `account_currency`, the nested `billing_address`, `description`, and `attrs` that hyperswitch's `build_payload_payment_request_data` emits.

## Fix

- Switch the connector amount converter and all **request** structs (`PayloadCardsRequestData`, `PayloadMandateRequestData`, `PayloadCardTokenRequestData`, `PayloadRefundRequest`, `PayloadClientAuthPayment`) from `FloatMajorUnit` to `StringMajorUnit`. **Response** structs keep `FloatMajorUnit` — the connector returns numeric amounts and RouterData parity was already clean.
- Add `account_class` (derived from `bank_holder_type`) and `account_currency` (`currency.to_string()`) to `PayloadBankAccountInner`; build the nested `billing_address` (mirroring the existing card builder); populate `description` (from `resource_common_data.description`) and `attrs` (filtered merchant metadata via a `get_filtered_metadata` helper mirroring hyperswitch). The card authorize path is unchanged (`description`/`attrs` stay `None`, omitted via `skip_serializing_if`).

## Verification

- `cargo check -p connector-integration` ✅
- `cargo clippy -p connector-integration` ✅
- Request builder now mirrors the hyperswitch `payload` reference field-for-field; final parity to be confirmed by a shadow-validation re-run.

Fixes juspay/hyperswitch-cloud#16753
Fixes juspay/hyperswitch-cloud#16754

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by XyneSpaces — COMMENTED** [review]
## Review Summary

**Verdict:** :white_check_mark: **Approve**

:rotating_light: 0 critical · :warning: 0 warning · :bulb: 3 suggestions · :mag: 0 nitpick

**Classification:** connector-bugfix-webhook · secondary: connector-flow-addition, tests-specs-docs
**Risk:** Medium

This PR fixes request parity issues between UCS and hyperswitch for the Payload connector by:
1. Switching amount type from `FloatMajorUnit` to `StringMajorUnit` (proper 2-decimal string serialization)
2. Adding missing ACH bank_debit fields: `account_class`, `account_currency`, `billing_address`, `description`, `attrs`
3. Updating field probe test fixtures to match

The changes align the connector with hyperswitch reference behavior as validated by shadow testing. Response handling correctly maintains `FloatMajorUnit` since Payload returns numeric amounts.

File: crates/integrations/connector-integration/src/connectors/payload/transformers.rs [hunk 1]
```diff
@@ -128,11 +128,28 @@ impl PayloadAuthType {
     }
 }
 
+/// Strip the connector-internal `processing_account_id` key from the merchant
+/// metadata before it is echoed back to Payload as `attrs` (mirrors hyperswitch).
+fn get_filtered_metadata(metadata: Option<&serde_json::Value>) -> Option<serde_json::Value> {
```
Comment (XyneSpaces) [inline crates/integrations/connector-integration/src/connectors/payload/transformers.rs:133]: :bulb: Consider documenting the `processing_account_id` exclusion rationale

The `get_filtered_metadata()` helper strips `processing_account_id` from metadata before echoing it back as `attrs`. Add a brief comment explaining why this field is excluded (e.g., "internal connector-service identifier that should not be forwarded to the provider").

This helps future maintainers understand the filtering logic.

File: crates/integrations/connector-integration/src/connectors/payload/requests.rs [hunk 2]
```diff
@@ -84,9 +90,18 @@ pub struct PayloadBank {
 
 #[derive(Clone, Debug, Serialize, PartialEq)]
 pub struct PayloadBankAccountInner {
+    pub account_class: Option<PayloadAccClass>,
+    pub account_currency: String,
     pub account_number: Secret<String>,
-    pub routing_number: Secret<String>,
     pub account_type: PayloadBankAccountType,
```
Comment (XyneSpaces) [inline crates/integrations/connector-integration/src/connectors/payload/requests.rs:96]: :bulb: Document the `PayloadAccClass` enum purpose

Add a brief doc comment explaining that this maps to the ACH bank account holder type (Personal/Business) for the Payload API. This improves discoverability for developers unfamiliar with Payload's account classification scheme.

Example:
```rust
/// Account classification for ACH bank transfers
/// Maps to `bank_holder_type` values in the domain model
```

File: crates/integrations/connector-integration/src/connectors/payload/transformers.rs [hunk 3]
```diff
@@ -264,13 +316,15 @@ fn build_payload_bank_account_request_data<T: PaymentMethodDataTypes>(
                 amount,
```
Comment (XyneSpaces) [inline crates/integrations/connector-integration/src/connectors/payload/transformers.rs:328]: :bulb: Consider adding unit test for `get_filtered_metadata` edge cases

The `get_filtered_metadata` helper handles three cases (Object with key, Object without key, non-Object). Consider adding a unit test verifying:
1. `processing_account_id` is stripped from nested objects
2. Empty object after filtering returns `None`
3. Non-object metadata returns `None`

This ensures the filtering logic remains correct through future refactors.

**Review by AkshayaFoiger — DISMISSED** [review]

File: crates/integrations/connector-integration/src/connectors/payload/transformers.rs [hunk 4]
```diff
@@ -128,11 +128,28 @@ impl PayloadAuthType {
     }
 }
 
+/// Strip the connector-internal `processing_account_id` key from the merchant
+/// metadata before it is echoed back to Payload as `attrs` (mirrors hyperswitch).
+fn get_filtered_metadata(metadata: Option<&serde_json::Value>) -> Option<serde_json::Value> {
+    metadata.and_then(|m| match m {
+        serde_json::Value::Object(map) => {
+            let mut filtered = map.clone();
+            filtered.remove("processing_account_id");
```
Comment (AmitsinghTanwar007) [inline crates/integrations/connector-integration/src/connectors/payload/transformers.rs:137]: why this fields comes in metadata 
and instead of filtering not required fields, you should choose a logic of extracting required values only.

Comment (shuklatushar226) [reply crates/integrations/connector-integration/src/connectors/payload/transformers.rs:137]: Good question — two parts:

**Why `processing_account_id` shows up in metadata:** Payload needs a `processing_id` (the merchant's processing account). Hyperswitch lets a merchant override it via connector metadata, so it reads `processing_account_id` from `metadata` first and falls back to the auth config — i.e. that key is consumed separately as the top-level `processing_id`, so it shouldn't also be echoed back inside `attrs`.

**Why exclude rather than extract an allow-list:** `attrs` is a free-form, merchant-defined bag — Payload stores it and returns it verbatim, and the merchant can put arbitrary keys in it (in this capture: `udf1`, `login_date`, `new_customer`). There's no fixed set of "required" keys to extract; the only field that needs special handling is `processing_account_id`. So the correct shape is "forward everything except the internal id."

It also has to match hyperswitch byte-for-byte — closing that request-parity diff is the whole purpose of this PR, and hyperswitch's own `get_filtered_metadata` does exactly this exclusion. An allow-list would drop merchant keys that hyperswitch forwards and re-open the diff.

Happy to add a doc comment capturing this rationale if that helps.

**Review by AmitsinghTanwar007 — DISMISSED** [review]

**Review by AmitsinghTanwar007 — APPROVED** [review]

**Review by kanikac199 — APPROVED** [review]
